src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -262,32 +276,27 @@ private void EnqueueWork(bool highPriority, bool frozenPartialSemantics, Cancell
                 // bails gracefully by checking below.
                 await _visibilityTracker.DelayWhileNonVisibleAsync(
                     _dataSource.ThreadingContext, _dataSource.AsyncListener, _subjectBuffer, DelayTimeSpan.NonFocus, cancellationToken).NoThrowAwaitable(captureContext: true);
-
-                if (cancellationToken.IsCancellationRequested)
-                    return default;
             }
 
-            _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
-            if (cancellationToken.IsCancellationRequested)
-                return default;
-
             using (Logger.LogBlock(FunctionId.Tagger_TagSource_RecomputeTags, cancellationToken))
             {
+                _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
+                if (cancellationToken.IsCancellationRequested)
+                    return null;
+
                 // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
                 // thread to do the computation. Finally, once new tags have been computed, then we update our state
                 // again on the foreground.
                 var spansToTag = GetSpansAndDocumentsToTag();
                 var caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
-                var oldTagTrees = this.CachedTagTrees;
-                var oldState = this.State;
-
-                var textChangeRange = this.AccumulatedTextChanges;
-                var subjectBufferVersion = _subjectBuffer.CurrentSnapshot.Version.VersionNumber;
 
-                await TaskScheduler.Default;
+                // If we're being called from within a blocking JTF.Run call, we don't want to switch to the background
+                // if we can avoid it.
+                if (!calledFromJtfRun)
```

Comment:
>calledFromJtfRun [](http://example.com/codeflow?start=21&length=16)

I don't suppose the highPriority flag could be used for this check instead?

Reply:
no.  The first call to get tags is also high-pri.